### PR TITLE
Enhance align actions

### PR DESCRIPTION
**Pull request by Ribakys**

### CODE REVIEW

**Comment by jrmi** [comment]
Hi @Ribakys ,

You need to rebase your work as I told you in the previous PR. f you need help on this point, feel free to ask me.

Thanks again for your work.

**Comment by Ribakys** [comment]
> 
> 
> Hi @Ribakys ,
> 
> You need to rebase your work as I told you in the previous PR. f you need help on this point, feel free to ask me.
> 
> Thanks again for your work.

Hi @jrmi ,
How exactly should I do?

**Comment by jrmi** [comment]
Read the whole comment before doing anything, I propose multiple options.

You can follow this procedure https://gist.github.com/ravibhure/a7e0918ff4937c9ea1c456698dcd58aa or this one https://medium.com/@topspinj/how-to-git-rebase-into-a-forked-repo-c9f05e821c8a

In short:
- Add my repository as a remote `git add upstream git@github.com:jrmi/airboardgame.git`
- Fetch the new remote `git fetch upstream`
- Create a branch on the current master to easily revert the process if something goes wrong `git branch savepoint`
- Rebase your master on the new remote master `git rebase upstream/master`
- Resolve the conflicts
- and force update your repository `git push --force`

You should have conflict difficult to resolve. My advice if it's too hard is to save the modified file in another directory then hard reset the master to the upstream master `git reset upstream master --hard`. You'll loose all your modifications (there still are in the git history though) but you start from fresh.

Remember that it's really hard to lose something with git. If you feel lost, stop everything and just ask me.

Another option if you don't feel confident: give me write access to your repository and I'll do it for you ;-)

**Comment by jrmi** [comment]
Hey @Ribakys , do you need help on this ?

**Comment by Ribakys** [comment]
I have completed the steps, but there seems to be a problem. I gave you access to my repository to simplify the resolution.

**Comment by jrmi** [comment]
> I have completed the steps, but there seems to be a problem. I gave you access to my repository to simplify the resolution.

I think it's OK ! If the CI complains now it's because secret variables are not accessible from fork for security reasons. (I've to migrate to github actions to solve this ;-) )

I'll review your work as soon as possible, thanks for your contribution!

**Comment by Ribakys** [comment]
> 
> 
> > I have completed the steps, but there seems to be a problem. I gave you access to my repository to simplify the resolution.
> 
> I think it's OK ! If the CI complains now it's because secret variables are not accessible from fork for security reasons. (I've to migrate to github actions to solve this ;-) )
> 
> I'll review your work as soon as possible, thanks for your contribution!

I make a video to test new functionalities :https://youtu.be/-euXkaM3p5U

File: src/components/boardComponents/useItemActions.jsx [hunk 1]
```diff
@@ -109,18 +109,15 @@ export const useItemActions = () => {
         )
       );
 
-      const [newX, newY] = [
-        (minMax.min.x + minMax.max.x) / 2,
-        (minMax.min.y + minMax.max.y) / 2,
-      ];
+      const [newX, newY] = [minMax.min.x, minMax.min.y];
```
Comment (jrmi) [inline src/components/boardComponents/useItemActions.jsx:112]: What seems odd to me with the new version:
- if you stack distant items, you can have all your items stacked to the top left of selection instead of middle of the selection. I like the "gathering" effect it give when grouping to the center.
- I think it less looks like a stack with the descendant order
- We lost stack elevation precision on little number of stacked item

My guess, by seeing your video, is that you're trying to have predictible and stable card alignement ? If you stack then align then stack again with the previous version, the cards are stacked in the center of the line which is not really convenient. My proposition: stack each item from the position of the deeper element in the global element array, the one which is "under" all other selected elements, this way if you press the same alignement multiple times, the cards doesn't move and if you change alignment then come back to the first one the cards get back to their original position. Is that ok for you ? In this case, you should modify your other alignment function to start from the same element.

I think you also want to avoid too high stack. For this point I propose to decrease the index modification by some factor (0.5) if we have more than, let's say, 32 cards in the stack ? What do you think ?

File: src/components/boardComponents/useItemActions.jsx [hunk 2]
```diff
@@ -132,6 +129,9 @@ export const useItemActions = () => {
     ({ snapshot }) => async () => {
```
Comment (jrmi) [inline src/components/boardComponents/useItemActions.jsx:129]: I think you can have the same effect but also deal with different sized item with something like this version:

```js
   ({ snapshot }) => async () => {
      const selectedItemList = await getSelectedItemList(snapshot);

      // Count number of elements
      const numberOfElements = selectedItemList.length;

      // Compute
      const minMax = { min: {}, max: {} };
      minMax.min.x = Math.min(...selectedItemList.map(({ x }) => x));
@@ -141,13 +141,13 @@ export const useItemActions = () => {
          ({ x, id }) => x + document.getElementById(id).clientWidth
        )
      );
      minMax.max.y = Math.max(...selectedItemList.map(({ y, id }) => y));

      // Changes starts here --------------
      let newX = minMax.min.x;

      batchUpdateItems(selectedItems, (item) => {
        const { clientWidth } = document.getElementById(item.id);
        const newItem = {
          ...item,
          x: newX,
          y: minMax.min.y,
        };
        newX += clientWidth
        return newItem;
      });
     // Changes ends here -----------
    },
    [getSelectedItemList, batchUpdateItems, selectedItems]
);
```

**Review by jrmi — CHANGES_REQUESTED** [review]

**Comment by jrmi** [comment]
You can rename the `align` function to `stack` if you want to for more consistency (was a bad naming).

**Comment by jrmi** [comment]
Hi @Ribakys, how it's going ?

**Comment by Ribakys** [comment]
As the two stacks (one centered and the other the top left) are useful to us, why not offer both. And then, if necessary, make other stacks.

Yes, why not keep the stack with the visual of ascending order.

Ok, we can test this criterion to improve the visual of the rise of the stack with 32 cards as a threshold.

I had already committed for factor 0.5: https://github.com/jrmi/airboardgame/pull/282/commits/6c93efa2ede97ec81836aa66335b7624f549cd49

And I had already committed on the stability of the alignments. Now they don't move.

I can rename the function 'align' to 'stackToCenter' and 'stackToTopLeft', what do you think?

To do with objects of different sizes, I will have to look.

**Comment by jrmi** [comment]
> As the two stacks (one centered and the other the top left) are useful to us, why not offer both. And then, if necessary, make other stacks.

I don't think that multiplying the same action with such minor difference is a good idea. What doesn't works for you with my proposition of aligning from the deeper element? May be if you explain me what you need that we can found a common solution (or may be not, in this case we'll keep the two versions)?

**Comment by Ribakys** [comment]
> 
> 
> > As the two stacks (one centered and the other the top left) are useful to us, why not offer both. And then, if necessary, make other stacks.
> 
> I don't think that multiplying the same action with such minor difference is a good idea. What doesn't works for you with my proposition of aligning from the deeper element (I think its even a better solution over the center alignement)? May be if you explain me what you need that we can found a common solution (or may be not, in this case we'll keep the two versions)?

My need was to maintain the position of the stack after a stack, alignment, or square arrangement.

For example, when creating a new card game, it is useful to be able to quickly arrange the imported cards. You may want to align or square a large number of cards. When you stack in the middle, the problem is that the stack is off the board. You then have to go get it and reposition it.

Even in-game, there are features that do not yet exist and that can be achieved through this. For example, quickly displaying the cards that are in a pile, then putting them back in the pile.

**Comment by jrmi** [comment]
> 
> My need was to maintain the position of the stack after a stack, alignment, or square arrangement.
> 
> For example, when creating a new card game, it is useful to be able to quickly arrange the imported cards. You may want to align or square a large number of cards. When you stack in the middle, the problem is that the stack is off the board. You then have to go get it and reposition it.
> 
> Even in-game, there are features that do not yet exist and that can be achieved through this. For example, quickly displaying the cards that are in a pile, then putting them back in the pile.

That was my guest, so what's not working with my last proposition?

**Comment by Ribakys** [comment]
> 
> 
> > My need was to maintain the position of the stack after a stack, alignment, or square arrangement.
> > For example, when creating a new card game, it is useful to be able to quickly arrange the imported cards. You may want to align or square a large number of cards. When you stack in the middle, the problem is that the stack is off the board. You then have to go get it and reposition it.
> > Even in-game, there are features that do not yet exist and that can be achieved through this. For example, quickly displaying the cards that are in a pile, then putting them back in the pile.
> 
> That was my guess and it's a good idea, so what's not working with my last proposition?

I think I misunderstood your proposal. Could you rephrase it please?

**Comment by jrmi** [comment]
I think we can keep the stability by aligning all items relatively to the deeper (z-order) element (should the first you found in the item id list). Its also more convenient if you put back card from discard onto the deck.

File: src/components/boardComponents/useItemActions.jsx [hunk 3]
```diff
@@ -89,8 +90,49 @@ export const useItemActions = () => {
...
+      const { clientWidth, clientHeight } = document.getElementById(item.id);
+      return {
+        ...item,
+        x: newX - clientWidth / 2 + index * stackThickness,
+        y: newY - clientHeight / 2 + (selectedItemList.length - index - 1) * stackThickness,
+      };
+    });
+  },
+  [getSelectedItemList, batchUpdateItems, selectedItems]
+);
```
Comment (jrmi) [inline src/components/boardComponents/useItemActions.jsx:132]: To complete my last comment, this is the partial fix I propose:

```suggestion
const stackToCenter = useRecoilCallback(
  ({ snapshot }) => async () => {
    const selectedItemList = await getSelectedItemList(snapshot);

    // Should be the center here -> the first card should not move!
    // Can be the exact same computation for other align actions
    const [newX, newY] = [selectedItemList[0].x, selectedItemList[0].y];
    
    let stackThickness = 0;
    if (selectedItemList.length >= 32) {
        stackThickness = 0.5;
    } else {
        stackThickness = 1;
    }
    let index = -1;
    batchUpdateItems(selectedItems, (item) => {
      index += 1;
      const { clientWidth, clientHeight } = document.getElementById(item.id);
      return {
        ...item,
        x: newX - clientWidth / 2 + index * stackThickness,
        y: newY - clientHeight / 2 + (selectedItemList.length - index - 1) * stackThickness,
      };
    });
  },
  [getSelectedItemList, batchUpdateItems, selectedItems]
);
```

Comment (jrmi) [reply src/components/boardComponents/useItemActions.jsx:132]: I think we can use the same origin item for all alignments, this way you keep stability and its seems more intuitive to me.
 
- For the stack, all cards are stacked on top of the first one.
- For the line the first item in `selectedItemList` is the most at the left then the second just after, ... it's like "carressing" a deck of card to allow payer to pick one (Like in this video https://youtu.be/0ytlTLoKknU?t=121)
- For the square, the same, the first item in the `selectedItemList` is the top left card then the second just after, etc...

If you switch from one alignement to another, the first card is the same and never move, so we have the stability even if you click multiple times

One use case, for example, is when you want to put back some cards in a deck you select all card and click stack action. The result should be that only previously discarded cards move and the other stays.

Comment (Ribakys) [reply src/components/boardComponents/useItemActions.jsx:132]: Ok, it's clearer to me, I understand now.

Comment (Ribakys) [reply src/components/boardComponents/useItemActions.jsx]: Effectively by using the first element of the list to carry out the stacks and the alignments, it is simpler and clearer.
I stabilized the displacement effect of the stack to center function.

I made the changes here: https://github.com/jrmi/airboardgame/pull/282/commits/8353a0b71b06dab64d3b5571e3e6c85eaa0adc1a

A video to illustrate:
https://youtu.be/_FUZJrGg5PM


I noticed two bugs that I show in two other videos:
https://youtu.be/nuk4UJTm7zY
https://youtu.be/ALtS6JRxhaE

Comment (jrmi) [reply src/components/boardComponents/useItemActions.jsx]: Nice, it's really cool. Unless you need it, you can remove the stack to center action to keep only yours now.

For the first bug, you have to know that the order of the cards in the item stack is reversed when you reveal the card to simulate the effect of having multiple cards in hand and reveal them: the one on bottom is the first you see. So the alignment is now from the one which was on top before. I don't think it should be a problem thought, but as I see in your video, if you click again it should keep stability and it doesn't... we have to investigate. 

For the second bug, I already noticed it but don't succeed to reproduce it in a predictive way. Your video will help me to do it, thanks. I already have a clue of where I need to look for.

**Comment by jrmi** [comment]
Also, can you rebase your work to benefit from new CI (I hope it will work now with cloned repository).

File: src/components/boardComponents/useItemActions.jsx [hunk 4]
```diff
@@ -89,70 +90,120 @@ export const useItemActions = () => {
...
-        )
-      );
-      minMax.max.y = Math.max(...selectedItemList.map(({ y }) => y));
+      // Count number of elements
+      const numberOfElements = selectedItemList.length;
 
-      const [newX, newY] = [minMax.min.x, minMax.max.y];
-      let index = -1;
+      let newX = selectedItemList[0].x;
+      let newY = selectedItemList[0].y;
```
Comment (jrmi) [inline src/components/boardComponents/useItemActions.jsx:194]:
```suggestion
      let { x:newX, y:newY} = selectedItemList[0];
```

File: src/components/boardComponents/useItemActions.jsx [hunk 5]
```diff
@@ -89,70 +90,120 @@ export const useItemActions = () => {
...
+      let index = numberOfElements;
+      let gapBetweenItems = 5; // Negative value is possible.
```
Comment (jrmi) [inline src/components/boardComponents/useItemActions.jsx:196]: Can you make it a function parameter with a default value? I plan to add parametrized actions (to avoid duplicate rotation action for each angle for example).

**Review by jrmi — CHANGES_REQUESTED** [review]

**Comment by jrmi** [comment]
Nice work, I can't wait to see it live.

**Comment by jrmi** [comment]
Actually I think I already fix the "item cloned" issue. Can you try to reproduce it after the rebase?

File: src/components/boardComponents/useItemActions.jsx (hunk 3 above)
Comment (Ribakys) [reply src/components/boardComponents/useItemActions.jsx]: Note that this bug only occurs with the "Reveal / Return" action and not with the "Return for me" action.
I did not understand why reverse the order?
Is it really necessary? Because it could be a source of bugs.

For the 2nd bug, I explain in the description of the video how to reproduce it.


For the "stackToCenter" action, are you sure you don't want to keep it?
If you don't want to, then here are some suggestions:
- we could keep the function in "useItemActions.jsx" (as it is coded, could it be used one day?)
- we keep the name of the function which is explicit "stackToTopLeft",
- you can change the "Stack To Top Left" label to "Stack".

**Comment by Ribakys** [comment]
> 
> 
> Actually I think I already fix the "item cloned" issue. Can you try to reproduce it after the rebase?

In Visual Studio Code, I do a "Pull (Rebase)" on "master" branch, but I don't know if this rebase are functioning.

**Comment by Ribakys** [comment]
> 
> 
> > Actually I think I already fix the "item cloned" issue. Can you try to reproduce it after the rebase?
> 
> In Visual Studio Code, I do a "Pull (Rebase)" on "master" branch, but I don't know if this rebase are functioning.

I did the rebase by hand with the instructions you gave me, and success !

Regarding the bug, it no longer appears: the fix is ok. And besides, the boxes are not already checked.

File: src/components/boardComponents/useItemActions.jsx (hunk 3 above)
Comment (jrmi) [reply src/components/boardComponents/useItemActions.jsx]: > Note that this bug only occurs with the "Reveal / Return" action and not with the "Return for me" action.
> I did not understand why reverse the order?
> Is it really necessary? Because it could be a source of bugs.

Sometimes in some games you return many cards at same times and their should be reversed. For example, in Solitaire games, you return the cards 3 by 3 and only the most on the top is available until you place it. Actually, that's the natural behaviour when you have real cards in your hands and I want that Airboargame mimic the natural behaviour as much as possible to be able to support real board games without rules or behaviours modification.

However, If you have a specific use case in mind where it can cause trouble, share it and we'll consider adding a non reversing reveal action (by adding a parameter to the function). The more we discuss, the more I think the ticket #317 should solve a lot of use case (but also make game creation a bit more complex ;-) ).

What do you think?

Comment (jrmi) [reply src/components/boardComponents/useItemActions.jsx]: > For the "stackToCenter" action, are you sure you don't want to keep it?
> If you don't want to, then here are some suggestions:
> 
>     * we could keep the function in "useItemActions.jsx" (as it is coded, could it be used one day?)
> 
>     * we keep the name of the function which is explicit "stackToTopLeft",
> 
>     * you can change the "Stack To Top Left" label to "Stack".

Ok with that!

Comment (Ribakys) [reply src/components/boardComponents/useItemActions.jsx]: > 
> 
> > Note that this bug only occurs with the "Reveal / Return" action and not with the "Return for me" action.
> > I did not understand why reverse the order?
> > Is it really necessary? Because it could be a source of bugs.
> 
> Sometimes in some games you return many cards at same times and their should be reversed. For example, in Solitaire games, you return the cards 3 by 3 and only the most on the top is available until you place it. Actually, that's the natural behaviour when you have real cards in your hands and I want that Airboargame mimic the natural behaviour as much as possible to be able to support real board games without rules or behaviours modification.
> 
> However, If you have a specific use case in mind where it can cause trouble, share it and we'll consider adding a non reversing reveal action (by adding a parameter to the function). The more we discuss, the more I think the ticket #317 should solve a lot of use case (but also make game creation a bit more complex ;-) ).
> 
> What do you think?

I think there are many use case when also we need to keep order. For example, when we want to return our cards in hand, the order don't change.

The both case can be natural :
- by rotation of the hand,  we can horizontally return  several cards at the same time : the order are inverted.
- by rotation of the hand, we can vertically return several cards at the same time: the order are not inverted.

 The ticket #317 should  make game creation a bit more complex (more possibilities), but not necessary difficult :-p

Comment (jrmi) [reply src/components/boardComponents/useItemActions.jsx]: Fine, keep both actions ;-)

File: src/components/boardComponents/index.jsx [hunk 6]
```diff
@@ -35,7 +35,7 @@ export const itemMap = {
     component: Rect,
     defaultActions: ["lock", "remove"],
     availableActions: [
-      "stack",
+      "stackToTopLeft",
```
Comment (jrmi) [inline src/components/boardComponents/index.jsx:38]: I think you must keep the name "stack" to keep compatibility with already existing games.

File: src/components/boardComponents/useItemActions.jsx (hunk 3 above)
Comment (jrmi) [reply src/components/boardComponents/useItemActions.jsx]: You also can do it in another PR if you prefer to have this PR live earlier.

**Comment by jrmi** [comment]
Ping ?

**Comment by Ribakys** [comment]
> 
> 
> Just to sum up:
> 
>     * If you rename action "stackToTopLeft" to "stack" to keep compatibility for old games
--> It's ok.

> 
>     * and if you add the alignement gap as a parameter
--> I don't understand.

> 
> 
> The PR can be validated.
> 
> Tell me if it's ok for you.

**Comment by jrmi** [comment]
> > ```
> > * and if you add the alignement gap as a parameter
> > ```
> 
> --> I don't understand.
> 

You did it ;-)

I have some little fix and I need to rebase. Do you allow me to push the result on your repository to keep this PR and merge it ? Otherwise I have to create another one.

**Comment by Ribakys** [comment]
> 
> 
> > > ```
> > > * and if you add the alignement gap as a parameter
> > > ```
> > 
> > 
> > --> I don't understand.
> 
> You did it ;-)
> 
> I have some little fixes and I need to rebase. Do you allow me to push the result on your repository to keep this PR and merge it ? Otherwise I have close this one and create another one.

Ok. Yes you can push the result on my repository.

**Review by jrmi — APPROVED** [review]
